### DISCUSSION TITLE
Key-event notification

### DISCUSSION THREAD

**Discussion by sebysr** [NodeRedis/node_redis#2792] [Q&A]
Hi all, I want to know if anyone has faced the following issue.
I have a nodejs application with a redis client with a key-event subscription and when my keys are expired i manage it with some actions. In my case, I've two pods of the nodejs application, so i've 2 notifications for each expired key.
How I can manage this kind of situation and manage only once time the expiration?
Have you some suggestion?
Thanks in advance.

**Comment by Pleo2** [comment]
What you are seeing is expected: keyspace notifications use Redis Pub/Sub, so every subscribed pod receives the expiration event. Pub/Sub does not elect one consumer, persist the event, or provide acknowledgement/retry semantics.

If the action is best-effort, each pod can attempt an atomic claim before doing the work, for example `SET expiration-claim:<unique-job-id> <worker-id> NX EX <ttl>`. Only the pod receiving `OK` proceeds. Make the handler idempotent, and use a unique job/version identifier rather than only the Redis key name if that key can be recreated.

If the action must run reliably exactly once from the application's perspective, do not use an expiration notification as the work queue. Redis documents Pub/Sub delivery as **at-most-once**: an event is lost if every subscriber is disconnected when it is published. A stronger design is:

1. Store the deadline in a sorted set (or create a delayed-job record).
2. Have workers atomically claim due jobs.
3. Put claimed work into a Redis Stream consumer group or another durable queue.
4. Acknowledge only after an idempotent handler succeeds.

A single dedicated subscriber removes duplicate processing between pods, but it still does not solve lost events during disconnects or restarts.

References: [keyspace notifications](https://redis.io/docs/latest/develop/pubsub/keyspace-notifications/) and [Pub/Sub delivery semantics](https://redis.io/docs/latest/develop/pubsub/#delivery-semantics).
